### ISSUE TITLE
MongoDB not respecting limits

### ISSUE THREAD

**Issue by AleCo3lho** [mongodb/mongodb-kubernetes-operator#936]
Hello guys, I have a MongoDB Operator deployed in HA, and I'm facing this problem that even limiting the amount of memory that mongod must use, this limit is not being respected, which ends up taking a high memory consumption and consequently the pod is dropped.

The node itself has 32gb of ram and we are reserving 30gb for the mongod container and limiting consumption to 23GB

I have here some information from Grafana, showing the consumption of the container before a failure.

![memory](https://user-images.githubusercontent.com/20954739/161768821-f190f044-8fe2-432f-ac6d-c047db488f42.png)

Just like the mongodb CRD yaml.

```yaml
apiVersion: v1
items:
- apiVersion: mongodbcommunity.mongodb.com/v1
  kind: MongoDBCommunity
  metadata:
    annotations:
      mongodb.com/v1.lastAppliedMongoDBVersion: 5.0.6
    generation: 9
    name: prd-mongodb
    namespace: shared-database
  spec:
    additionalMongodConfig:
      storage.wiredTiger.engineConfig.cacheSizeGB: 23
      storage.wiredTiger.engineConfig.journalCompressor: zlib
    members: 3
    security:
      authentication:
        modes:
        - SCRAM
    statefulSet:
      spec:
        template:
          spec:
            containers:
            - env:
              - name: MANAGED_SECURITY_CONTEX
                value: "true"
              name: mongod
              resources:
                limits:
                  cpu: "3.5"
                  memory: 30Gi
                requests:
                  cpu: "2"
                  memory: 14Gi
            - env:
              - name: MANAGED_SECURITY_CONTEX
                value: "true"
              name: mongodb-agent
              resources:
                limits:
                  cpu: "0.5"
                  memory: 512M
                requests:
                  cpu: "0.2"
                  memory: 200M
        volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 50Gi
            storageClassName: managed-premium
        - metadata:
            name: logs-volume
          spec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
            storageClassName: managed-premium
    type: ReplicaSet
    users:
    - db: admin
      name: prd-mongodb-user
      passwordSecretRef:
        name: prd-mongodb-user-password
      roles:
      - db: admin
        name: clusterAdmin
      - db: admin
        name: userAdminAnyDatabase
      - db: admin
        name: root
      scramCredentialsSecretName: my-scram
    version: 5.0.6
  status:
    currentMongoDBMembers: 3
    currentStatefulSetReplicas: 3
    mongoUri: <HIDE>
    phase: Running
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
Thank you so much everyone for your attention.

**Comment by xtianus79** [comment]
This shouldn't be closed?

**Comment by irajdeep** [comment]
Hi all! 

@LoonyRules, first of all, thank you for your detailed explanation. Apologies for not having replied to this issue earlier and closing it preemptively; the team is currently a bit constrained and hasn't been able to get to this. 

I will open this ticket and track this internally so that we can evaluate it against other priorities. 

Best,
Raj

**Comment by mntolia** [comment]
I am facing the same issue. I am reposting what I posted on the MongoDB forums:

> I use MongoDB v5.0.3 via docker on a single AWS instance. The container kept crashing on long queries exactly after hitting 100% memory.
> Then I noticed this:
> This is the docker stat. You can notice that the memory is limited to 4 GB.
> image
> image
> 1231×208 17.2 KB
> This is the MongoDB host info:
> 
> {
>   system: {
>     currentTime: 2022-11-28T10:23:25.772Z,
>     hostname: 'mongo',
>     cpuAddrSize: 64,
>     memSizeMB: Long("31304"),
>     memLimitMB: Long("31304"),
>     numCores: 8,
>     cpuArch: 'x86_64',
>     numaEnabled: false
>   }
> You can see that the memLimitMB is set to 31 GB.
> 
> I rebuilt the container several times, but it still has the same issue.

**Comment by tzorzan** [comment]
The only reasonable setup I found that works for me is:
- remove memory limit to avoid OOM killing 
- set the memory request with common sizing considerations
- set explicitly Wired Tiger size based on memory request
- let the Linux kernel and k8s do it’s work handling the page size

I found this setup pretty good for my situation, dedicating almost a full node for MongoDB, and others pod should be “good citizens” setting their own requests and limits.

Best,
T

**Comment by gmartynov** [comment]
@irajdeep good day, this issue is still relevant, is there a hope to resolve it ?

**Comment by gmartynov** [comment]
up

**Comment by AleCo3lho** [comment]
Hey @slaskawi, I can't apply the same procedures you did, because we moved this specific client to a Persona implementation of the MongoDB, that seems to solve the problem.

I hope @gmartynov can help you with that, seems he is suffering from the same problem, I will keep watching these issue, seems its of my interest.

**Comment by slaskawi** [comment]
Thank @AleCo3lho !

I also skimmed through @LoonyRules's comment and just wanted to highlight, that autodetecting limits only works if there are no custom WT options specified (such as `storage.wiredTiger.engineConfig.cacheSizeGB`). If you override anything manually, the WT will respect your settings and getting them wrong (e.g. specifying more than 80% of the container capacity) might cause OOM kills. So in this case, the user needs to decide - rely on automation or specify settings manually.

**Comment by gmartynov** [comment]
Hello @slaskawi !
As I mentioned in similar issue https://github.com/mongodb/mongodb-kubernetes-operator/issues/830 we meet this problem in k8s cluster with cgroupv2. 
At the same time - in our another cluster with cgroupv1 everything is ok.

We use mongodb kubernetes operator installation.
Specs provided below, I removed strings autogenerated by k8s.

Mongodb community spec:
```
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: asrv-mongodb
  namespace: mongodb-operator
spec:
  additionalMongodConfig:
    setParameter:
      maxTransactionLockRequestTimeoutMillis: 10000
      transactionLifetimeLimitSeconds: 10800
    storage.wiredTiger.engineConfig.journalCompressor: zlib
  members: 3
  security:
    authentication:
      ignoreUnknownUsers: true
      modes:
      - SCRAM
  statefulSet:
    spec:
      template:
        spec:
          containers:
          - name: mongod
            resources:
              limits:
                cpu: 4
                memory: 6Gi
              requests:
                cpu: 1
                memory: 3Gi
      volumeClaimTemplates:
      - metadata:
          name: data-volume
        spec:
          accessModes:
          - ReadWriteOnce
          resources:
            requests:
              storage: 300Gi
          storageClassName: local-ssd-mongodata
      - metadata:
          name: logs-volume
        spec:
          accessModes:
          - ReadWriteOnce
          resources:
            requests:
              storage: 3Gi
          storageClassName: csi-cephfs-sc
  type: ReplicaSet
  users:
  - db: admin
    name: root
    passwordSecretRef:
      name: root-password
    roles:
    - db: admin
      name: root
    scramCredentialsSecretName: my-scram
  version: 4.2.17
```

Statefulset spec (generated by operator):
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app: asrv-mongodb-svc
  name: asrv-mongodb
  namespace: mongodb-operator
  ownerReferences:
  - apiVersion: mongodbcommunity.mongodb.com/v1
    blockOwnerDeletion: true
    controller: true
    kind: MongoDBCommunity
    name: asrv-mongodb
spec:
  podManagementPolicy: OrderedReady
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: asrv-mongodb-svc
  serviceName: asrv-mongodb-svc
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: asrv-mongodb-svc
    spec:
      containers:
      - command:
        - /bin/sh
        - -c
        - |2+

          #run post-start hook to handle version changes
          /hooks/version-upgrade

          # wait for config and keyfile to be created by the agent
           while ! [ -f /data/automation-mongod.conf -a -f /var/lib/mongodb-mms-automation/authentication/keyfile ]; do sleep 3 ; done ; sleep 2 ;

          # start mongod with this configuration
          exec mongod -f /data/automation-mongod.conf;

        env:
        - name: AGENT_STATUS_FILEPATH
          value: /healthstatus/agent-health-status.json
        image: docker.io/mongo:4.2.17
        imagePullPolicy: IfNotPresent
        name: mongod
        resources:
          limits:
            cpu: "4"
            memory: 6Gi
          requests:
            cpu: "1"
            memory: 3Gi
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /var/lib/mongodb-mms-automation/authentication
          name: asrv-mongodb-keyfile
        - mountPath: /data
          name: data-volume
        - mountPath: /healthstatus
          name: healthstatus
        - mountPath: /hooks
          name: hooks
        - mountPath: /var/log/mongodb-mms-automation
          name: logs-volume
        - mountPath: /tmp
          name: tmp
      - command:
        - /bin/bash
        - -c
        - |-
          current_uid=$(id -u)
          AGENT_API_KEY="$(cat /mongodb-automation/agent-api-key/agentApiKey)"
          declare -r current_uid
          if ! grep -q "${current_uid}" /etc/passwd ; then
          sed -e "s/^mongodb:/builder:/" /etc/passwd > /tmp/passwd
          echo "mongodb:x:$(id -u):$(id -g):,,,:/:/bin/bash" >> /tmp/passwd
          export NSS_WRAPPER_PASSWD=/tmp/passwd
          export LD_PRELOAD=libnss_wrapper.so
          export NSS_WRAPPER_GROUP=/etc/group
          fi
          agent/mongodb-agent -healthCheckFilePath=/var/log/mongodb-mms-automation/healthstatus/agent-health-status.json -serveStatusPort=5000 -cluster=/var/lib/automation/config/cluster-config.json -skipMongoStart -noDaemonize -useLocalMongoDbTools
        env:
        - name: AGENT_STATUS_FILEPATH
          value: /var/log/mongodb-mms-automation/healthstatus/agent-health-status.json
        - name: AUTOMATION_CONFIG_MAP
          value: asrv-mongodb-config
        - name: HEADLESS_AGENT
          value: "true"
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        image: quay.io/mongodb/mongodb-agent:12.0.10.7591-1
        imagePullPolicy: Always
        name: mongodb-agent
        readinessProbe:
          exec:
            command:
            - /opt/scripts/readinessprobe
          failureThreshold: 40
          initialDelaySeconds: 5
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 500M
          requests:
            cpu: 500m
            memory: 400M
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /opt/scripts
          name: agent-scripts
        - mountPath: /var/lib/mongodb-mms-automation/authentication
          name: asrv-mongodb-keyfile
        - mountPath: /var/lib/automation/config
          name: automation-config
          readOnly: true
        - mountPath: /data
          name: data-volume
        - mountPath: /var/log/mongodb-mms-automation/healthstatus
          name: healthstatus
        - mountPath: /var/log/mongodb-mms-automation
          name: logs-volume
        - mountPath: /tmp
          name: tmp
      dnsPolicy: ClusterFirst
      initContainers:
      - command:
        - cp
        - version-upgrade-hook
        - /hooks/version-upgrade
        image: quay.io/mongodb/mongodb-kubernetes-operator-version-upgrade-post-start-hook:1.0.5
        imagePullPolicy: Always
        name: mongod-posthook
        resources: {}
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /hooks
          name: hooks
      - command:
        - cp
        - /probes/readinessprobe
        - /opt/scripts/readinessprobe
        image: quay.io/mongodb/mongodb-kubernetes-readinessprobe:1.0.11
        imagePullPolicy: Always
        name: mongodb-agent-readinessprobe
        resources: {}
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /opt/scripts
          name: agent-scripts
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 2000
        runAsNonRoot: true
        runAsUser: 2000
      serviceAccount: mongodb-database
      serviceAccountName: mongodb-database
      terminationGracePeriodSeconds: 30
      volumes:
      - emptyDir: {}
        name: agent-scripts
      - emptyDir: {}
        name: asrv-mongodb-keyfile
      - name: automation-config
        secret:
          defaultMode: 416
          secretName: asrv-mongodb-config
      - emptyDir: {}
        name: healthstatus
      - emptyDir: {}
        name: hooks
      - emptyDir: {}
        name: tmp
  updateStrategy:
    type: RollingUpdate
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      creationTimestamp: null
      name: data-volume
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 300Gi
      storageClassName: local-ssd-mongodata
      volumeMode: Filesystem
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: logs-volume
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 3Gi
      storageClassName: csi-cephfs-sc
      volumeMode: Filesystem
```


Messages related to wired tiger in logs:
```
2023-08-10T15:37:55.545+0000 I  STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=63793M,cache_overflow=(file_max=0M),session_max=33000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=zlib),file_manager=(close_idle_time=100000,close_scan_interval=10,close_handle_minimum=250),statistics_log=(wait=0),verbose=[recovery_progress,checkpoint_progress],


2023-08-10T15:37:55.545+0000 I  CONTROL  [initandlisten] options: { config: "/data/automation-mongod.conf", net: { bindIp: "0.0.0.0", port: 27017 }, replication: { replSetName: "asrv-mongodb" }, security: { authorization: "enabled", keyFile: "/var/lib/mongodb-mms-automation/authentication/keyfile" }, setParameter: { authenticationMechanisms: "SCRAM-SHA-256", maxTransactionLockRequestTimeoutMillis: "10000", transactionLifetimeLimitSeconds: "10800" }, storage: { dbPath: "/data", engine: "wiredTiger", wiredTiger: { engineConfig: { journalCompressor: "zlib" } } } }
```

Grafana stats example on crash:
![image](https://github.com/mongodb/mongodb-kubernetes-operator/assets/30998648/34d22747-8531-410b-9b7a-4fb1c10e1202)

Workload stats example near the crash:
gmartyno@note-gmartyno-wsl:~$ kubectl top pod
NAME                                           CPU(cores)   MEMORY(bytes)
asrv-mongodb-0                                 4009m        6222Mi
asrv-mongodb-1                                 4006m        6185Mi
asrv-mongodb-2                                 47m          6230Mi
manage-0                                       27m          192Mi
mongodb-external-85d7bc7556-8hw7b              6m           83Mi
mongodb-kubernetes-operator-54b86cbcfd-xkz8m   4m           31Mi


As you see we don't modify `storage.wiredTiger.engineConfig.cacheSizeGB` parameter.

And in cgroupv2 there is another path to check memory metrics in the pod:
```
$ kubectl exec asrv-mongodb-0 -- cat /sys/fs/cgroup/memory.max
Defaulted container "mongod" out of: mongod, mongodb-agent, mongod-posthook (init), mongodb-agent-readinessprobe (init)
6442450944
```

Hope you will resolve this!

**Comment by akramincity** [comment]
I am facing a very weird problem no matter what i do i am unable to set resources requests and limits 
I can see the changes in mdbc but the statefulset is locked at default values

`apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: support-chat-mongodb
spec:
  members: 2
  type: ReplicaSet
  version: "7.0.0"
  replicaSetHorizons:
  - horizon: support-chat-mongodb.local.cm.dev:30010
  - horizon: support-chat-mongodb.local.cm.dev:30011

  security:
    authentication:
      modes:
      - SCRAM-SHA-1
      - SCRAM

  users:
    - name: root
      db: admin
      passwordSecretRef:
        name: mongo-password
      roles:
        - name: clusterAdmin
          db: admin
        - name: userAdminAnyDatabase
          db: admin
        - name: readWriteAnyDatabase
          db: admin
        - name: dbAdminAnyDatabase
          db: admin
      scramCredentialsSecretName: my-scram

  statefulSet:
    spec:
      serviceName: support-chat-mongodb-svc
      containers:
        - name: mongod
          resources:
            # limits:
            #   cpu: "4"
            #   memory: 32000M
            # requests:
            #   cpu: "3"
            #   memory: 28000M
          readinessProbe:
            failureThreshold: 50
            initialDelaySeconds: 10

      selector:
        matchLabels:
          db: mongodb

      volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: gp2
            resources:
              requests:
                storage: 300Gi

      template:
        metadata:
          labels:
            db: mongodb

        spec:
          affinity:
            podAntiAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: db
                        operator: In
                        values:
                          - mongodb
                  topologyKey: "kubernetes.io/hostname" 


          nodeSelector:
            karpenter.sh/provisioner-name: support-chat-mongo
          tolerations:
          - key: "support-chat-mongo"
            operator: "Exists"
            effect: "NoSchedule" `

**Comment by akramincity** [comment]
I am facing the problem that i am unable to set limits and requests of the statefulset i can see the changes are made in mdbc but when i describe my sts the values are not in effect
`apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: support-chat-mongodb
spec:
  members: 2
  type: ReplicaSet
  version: "7.0.0"
  replicaSetHorizons:
  - horizon: support-chat-mongodb.some.dev:30010
  - horizon: support-chat-mongodb.some.dev:30011

  security:
    authentication:
      modes:
      - SCRAM-SHA-1
      - SCRAM

  users:
    - name: root
      db: admin
      passwordSecretRef:
        name: mongo-password
      roles:
        - name: clusterAdmin
          db: admin
        - name: userAdminAnyDatabase
          db: admin
        - name: readWriteAnyDatabase
          db: admin
        - name: dbAdminAnyDatabase
          db: admin
      scramCredentialsSecretName: my-scram

  statefulSet:
    spec:
      serviceName: support-chat-mongodb-svc
      containers:
        - name: mongod
          resources:
            limits:
              cpu: "4"
              memory: 32000M
            requests:
              cpu: "3"
              memory: 28000M
          readinessProbe:
            failureThreshold: 50
            initialDelaySeconds: 10

      selector:
        matchLabels:
          db: mongodb

      volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: gp2
            resources:
              requests:
                storage: 300Gi

      template:
        metadata:
          labels:
            db: mongodb

        spec:
          affinity:
            podAntiAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: db
                        operator: In
                        values:
                          - mongodb
                  topologyKey: "kubernetes.io/hostname" 


          nodeSelector:
            karpenter.sh/provisioner-name: support-chat-mongo
          tolerations:
          - key: "support-chat-mongo"
            operator: "Exists"
            effect: "NoSchedule" `
          Adding operator yaml as well
`apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    email: support@mongodb.com
  labels:
    owner: mongodb
  name: mongodb-kubernetes-operator
spec:
  replicas: 1
  selector:
    matchLabels:
      name: mongodb-kubernetes-operator
  strategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      labels:
        name: mongodb-kubernetes-operator
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: name
                operator: In
                values:
                - mongodb-kubernetes-operator
            topologyKey: kubernetes.io/hostname
      nodeSelector:
          karpenter.sh/provisioner-name: support-chat-mongo
      tolerations:
      - key: "support-chat-mongo"
        operator: "Exists"
        effect: "NoSchedule"
      containers:
      - command:
        - /usr/local/bin/entrypoint
        env:
        - name: WATCH_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: OPERATOR_NAME
          value: mongodb-kubernetes-operator
        - name: AGENT_IMAGE
          value: quay.io/mongodb/mongodb-agent:12.0.25.7724-1
        - name: VERSION_UPGRADE_HOOK_IMAGE
          value: quay.io/mongodb/mongodb-kubernetes-operator-version-upgrade-post-start-hook:1.0.7
        - name: READINESS_PROBE_IMAGE
          value: quay.io/mongodb/mongodb-kubernetes-readinessprobe:1.0.15
        - name: MONGODB_IMAGE
          value: mongodb-community-server
        - name: MONGODB_REPO_URL
          value: docker.io/mongodb
        image: quay.io/mongodb/mongodb-kubernetes-operator:0.8.2
        imagePullPolicy: Always
        name: mongodb-kubernetes-operator
        resources:
          limits:
            cpu: 1100m
            memory: 1Gi
          requests:
            cpu: 500m
            memory: 200Mi
        securityContext:
          readOnlyRootFilesystem: true
          runAsUser: 2000
          allowPrivilegeEscalation: false
      securityContext:
        seccompProfile:
          type: RuntimeDefault
      serviceAccountName: mongodb-kubernetes-operator`

**Comment by akramincity** [comment]
I am unable to add limits and requests in mongod, it is set at default value but I can see the changes in mdbc  only but when i describe the sts values are changed to default 

`apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: support-chat-mongodb
spec:
  members: 2
  type: ReplicaSet
  version: "7.0.0"
  replicaSetHorizons:
  - horizon: support-chat-mongodb.some.dev:30010
  - horizon: support-chat-mongodb.some.dev:30011

  security:
    authentication:
      modes:
      - SCRAM-SHA-1
      - SCRAM

  users:
    - name: root
      db: admin
      passwordSecretRef:
        name: mongo-password
      roles:
        - name: clusterAdmin
          db: admin
        - name: userAdminAnyDatabase
          db: admin
        - name: readWriteAnyDatabase
          db: admin
        - name: dbAdminAnyDatabase
          db: admin
      scramCredentialsSecretName: my-scram

  statefulSet:
    spec:
      serviceName: support-chat-mongodb-svc
      containers:
        - name: mongod
          resources:
            limits:
              cpu: "4"
              memory: 32000M
            requests:
              cpu: "3"
              memory: 28000M
          readinessProbe:
            failureThreshold: 50
            initialDelaySeconds: 10

      selector:
        matchLabels:
          db: mongodb

      volumeClaimTemplates:
        - metadata:
            name: data-volume
          spec:
            accessModes: [ "ReadWriteOnce" ]
            storageClassName: gp2
            resources:
              requests:
                storage: 300Gi

      template:
        metadata:
          labels:
            db: mongodb

        spec:
          affinity:
            podAntiAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: db
                        operator: In
                        values:
                          - mongodb
                  topologyKey: "kubernetes.io/hostname" 


          nodeSelector:
            karpenter.sh/provisioner-name: support-chat-mongo
          tolerations:
          - key: "support-chat-mongo"
            operator: "Exists"
            effect: "NoSchedule" 
            ---
            apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    email: support@mongodb.com
  labels:
    owner: mongodb
  name: mongodb-kubernetes-operator
spec:
  replicas: 1
  selector:
    matchLabels:
      name: mongodb-kubernetes-operator
  strategy:
    rollingUpdate:
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      labels:
        name: mongodb-kubernetes-operator
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: name
                operator: In
                values:
                - mongodb-kubernetes-operator
            topologyKey: kubernetes.io/hostname
      nodeSelector:
          karpenter.sh/provisioner-name: support-chat-mongo
      tolerations:
      - key: "support-chat-mongo"
        operator: "Exists"
        effect: "NoSchedule"
      containers:
      - command:
        - /usr/local/bin/entrypoint
        env:
        - name: WATCH_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: OPERATOR_NAME
          value: mongodb-kubernetes-operator
        - name: AGENT_IMAGE
          value: quay.io/mongodb/mongodb-agent:12.0.25.7724-1
        - name: VERSION_UPGRADE_HOOK_IMAGE
          value: quay.io/mongodb/mongodb-kubernetes-operator-version-upgrade-post-start-hook:1.0.7
        - name: READINESS_PROBE_IMAGE
          value: quay.io/mongodb/mongodb-kubernetes-readinessprobe:1.0.15
        - name: MONGODB_IMAGE
          value: mongodb-community-server
        - name: MONGODB_REPO_URL
          value: docker.io/mongodb
        image: quay.io/mongodb/mongodb-kubernetes-operator:0.8.2
        imagePullPolicy: Always
        name: mongodb-kubernetes-operator
        resources:
          limits:
            cpu: 1100m
            memory: 1Gi
          requests:
            cpu: 500m
            memory: 200Mi
        securityContext:
          readOnlyRootFilesystem: true
          runAsUser: 2000
          allowPrivilegeEscalation: false
      securityContext:
        seccompProfile:
          type: RuntimeDefault
      serviceAccountName: mongodb-kubernetes-operator`

**Comment by captainhorsepower** [comment]
@gmartynov Hi. I've faced same issue with cgroups v2 today. And found this [jira ticket](https://jira.mongodb.org/browse/SERVER-60412). Turns out mongodb team added support for cgroups v2 and backported it to older versions starting with: 4.4.14, 5.3.0, 5.0.7, 4.2.20. 

My cluster was on 4.4.5 and `hostInfo.system.memLimitMB` was equal to `hostInfo.system.memSizeMB`. After updating to 4.4.14 (and later) it tolerates cgroups.

I noticed you used `docker.io/mongo:4.2.17` < 4.2.20 and 5.0.6 < 5.0.7. 

Hope this helps.
